_ext/eclipse-base/src/main/java/com/diffplug/spotless/extra/eclipse/base/service/SingleSlf4JService.java

```diff
@@ -0,0 +1,532 @@
+		consumer.accept(this);
+	}
+
+	@Override
+	public void audit(String message) {
+		log(new SimpleLogEntry(LogLevel.AUDIT, message));
+	}
+
+	@Override
+	public void audit(String format, Object arg) {
+		audit(String.format(format, arg));
+	}
+
+	@Override
+	public void audit(String format, Object arg1, Object arg2) {
+		audit(String.format(format, arg1, arg2));
+	}
+
+	@Override
+	public void audit(String format, Object... arguments) {
+		audit(String.format(format, arguments));
+	}
+
+	public static class SimpleLogEntry implements LogEntry {
+
+		private final LogLevel level;
+		private final String message;
+		private final Optional<Throwable> execption;
+
+		public SimpleLogEntry(LogLevel level, String message) {
+			this(level, message, Optional.empty());
+		}
+
+		public SimpleLogEntry(LogLevel level, String message, @Nullable Throwable execption) {
+			this(level, message, Optional.ofNullable(execption));
+		}
+
+		private SimpleLogEntry(LogLevel level, String message, Optional<Throwable> execption) {
+			this.level = level;
+			this.message = message;
+			this.execption = execption;
+		}
+
+		@Override
+		public Bundle getBundle() {
+			//Return the spotless framework bundle
+			return InternalPlatform.getDefault().getBundleContext().getBundle();
+		}
+
+		@SuppressWarnings({"rawtypes", "unchecked"})
+		@Override
+		public ServiceReference getServiceReference() {
+			return null;
+		}
+
+		@Override
+		@Deprecated
+		//Backward compatibility with Eclipse OSGI 3.12
+		public int getLevel() {
+			switch (level) {
+			case DEBUG:
+			case TRACE:
+				return SingleSlf4JService.LOG_DEBUG;
+			case AUDIT:
+			case INFO:
+				return SingleSlf4JService.LOG_INFO;
+			case ERROR:
+				return SingleSlf4JService.LOG_ERROR;
+			case WARN:
+				return SingleSlf4JService.LOG_WARNING;
+			}
+			return SingleSlf4JService.LOG_ERROR; //Don't fail here. Just log it as error. This is anyway just for debugging internal problems.
+		}
+
+		@Override
+		public String getMessage() {
+			return message;
+		}
+
+		@Override
+		public Throwable getException() {
+			return execption.orElse(null);
+		}
+
+		@Override
+		public long getTime() {
+			return 0;
+		}
+
+		@Override
+		public String toString() {
+			StringWriter result = new StringWriter();
+			result.write(message);
+			if (execption.isPresent()) {
+				result.write('\n');
+				result.write(execption.get().toString());
+				result.write('\n');
+				execption.get().printStackTrace(new PrintWriter(result));
+			}
+			return result.toString();
+		}
+
+		@Override
+		public LogLevel getLogLevel() {
+			return level;
+		}
+
+		@Override
+		public String getLoggerName() {
+			return this.getClass().getSimpleName();
+		}
+
+		@Override
+		public long getSequence() {
+			return 0;
+		}
+
+		@Override
+		public String getThreadInfo() {
+			return null;
+		}
+
+		@Override
+		public StackTraceElement getLocation() {
+			return null;
+		}
```

Comment:
Don't you have this information if the `execption` isn't `Optional.empty()`?

Reply:
I must apologize. I wrote the POC some time ago and when I finally found time to work on it again, I forgot that there was some documentation and cleanup to do...
My comments and reduction of visibility should clarify the issue. Please have a look...

Reply:
Much better. Thanks!